### PR TITLE
[Android] BMI support on Java external extension infrastructure.

### PR DESCRIPTION
**Pull request by DonnaWuDongxia**
This commit aimed to add BMI support to the external Java extension
infrastructure, including underline modules refining.
Details listed as follows:
1. Refine the BindingObject and EventTarget implmentation.
2. Add MessageHandler and MessageInfo to keep the compatibility
   with customized extension JS code based on 'jsStub' module.
   It means, registering the exposed API to MessageHandler instead
   of annotating them.
3. BMI support on both the 'jsStub' module and the Java side.
4. A invert color sample to test the BMI support feature.

BUG=XWALK-5602

### CODE REVIEW

**Comment by DonnaWuDongxia** [comment]
@halton Please help to review.

**Comment by DonnaWuDongxia** [comment]
image1

image2

**Comment by DonnaWuDongxia** [comment]
Pictures above showing the test case of BMI. Pass a image in JS side to Java by BMI, and invert the color in Java side, then return the image back to JS side by BMI.

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectAutoJS.java [hunk 1]
```diff
@@ -10,32 +10,26 @@
...
-
-    public void init(int instanceId, int objectId) {
-        myObjectId = objectId;
-        extInstanceId = instanceId;
+public class BindingObjectAutoJS extends BindingObject{
+
+    public Object handleMessage(MessageInfo info) {
+        Object result = null;
+        ReflectionHelper extReflect =
+            mInstanceHelper.getExtension().getReflection();
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectAutoJS.java]: nit, 8-space

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/MessageHandler.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2013 Intel Corporation. All rights reserved.
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/MessageHandler.java]: nit, 2015

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/MessageHandler.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package org.xwalk.app.runtime.extension;
+
+import android.content.Intent;
+import android.util.Log;
+
+import java.lang.reflect.*;
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/MessageHandler.java]: Can we explicit list interface want to use?

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/MessageInfo.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    private XWalkExtensionClient mExtension;
+    private int mInstanceId;
+    private String mJsName;
+    private String mCallbackId;
+    private String mObjectId;
+    private String mCmd;
+    private JSONArray mArgs;
+    private ByteBuffer mBinaryArgs;
+
+    private int  AlignedWith4Bytes(int length) {
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/MessageInfo.java]: nit, one space after `int`

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/ReflectionHelper.java [hunk 5]
```diff
@@ -335,22 +367,41 @@ boolean isInstance(Object obj) {
...
-                            targetObj, msg.getString("name"), msg.getJSONArray("args"));
+                    result = invokeMethod(ext, instanceId, targetObj,
+                                          memberName, args);
+                    break;
+                case "newInstance":
+                    BindingObject newObj = (BindingObject)(invokeMethod(ext,
+                                          instanceId, targetObj,
+                                          memberName, args));
+                    result = info.getInstanceHelper().addBindingObject(
+                                  info.getObjectId(), newObj);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/ReflectionHelper.java]: bad indent above.

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java [hunk 6]
```diff
@@ -67,15 +70,20 @@ public XWalkExtensionClient(String name, String jsApi, String[] entryPoints, XWa
...
-            if (mJsApi == null || mJsApi.length() == 0) {
-                Log.e("Extension-" + mName, "Can't generate JavaScript stub for this extension.");
-                return;
-            }
+          useJsStubGeneration = true;
+          mReflection = new ReflectionHelper(this.getClass());
+          mJsApi = new JsStubGenerator(mReflection).generate();
+          if (mJsApi == null || mJsApi.length() == 0) {
+            Log.e("Extension-" + mName, "Can't generate JavaScript stub for this extension.");
+            return;
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java]: 4-space indent

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java [hunk 7]
```diff
@@ -176,7 +182,9 @@ public XWalkExtensionBindingObjectStore getInstanceStore(int instanceID) {
      * @param message the message from JavaScript code.
      */
     public void onMessage(int extensionInstanceID, String message) {
-        handleMessage(extensionInstanceID, message);
+      if (useJsStubGeneration) {
+        getInstanceHelper(extensionInstanceID).handleMessage(message);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java]: indent

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java [hunk 8]
```diff
@@ -187,7 +195,11 @@ public void onMessage(int extensionInstanceID, String message) {
      * @param extensionInstanceID the ID of extension instance where the message came from.
      * @param message the binary message from JavaScript code.
      */
-    public void onBinaryMessage(int extensionInstanceID, byte[] message) {}
+    public void onBinaryMessage(int extensionInstanceID, byte[] message) {
+      if (useJsStubGeneration) {
+        getInstanceHelper(extensionInstanceID).handleMessage(message);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java]: indent

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java [hunk 9]
```diff
@@ -197,75 +209,28 @@ public void onBinaryMessage(int extensionInstanceID, byte[] message) {}
      * @param message the message from JavaScript code.
      */
     public String onSyncMessage(int extensionInstanceID, String message) {
-        Object result = handleMessage(extensionInstanceID, message);
-        return (result != null) ? ReflectionHelper.objToJSON(result): "";
+      Object result = null; 
+      if (useJsStubGeneration) {
+        result = getInstanceHelper(extensionInstanceID).handleMessage(message);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java]: indent

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java [hunk 10]
```diff
@@ -197,75 +209,28 @@ public void onBinaryMessage(int extensionInstanceID, byte[] message) {}
...
-            JSONObject msgOut = new JSONObject(); 
-            msgOut.put("cmd", "error");
-            msgOut.put("level", level);
-            msgOut.put("msg", msg);
-            postMessage(instanceId, msgOut.toString());
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
+    protected ExtensionInstanceHelper getInstanceHelper(int instanceId) {
+       return instanceHelpers.get(instanceId);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java]: indent

**Comment by halton** [comment]
LGTM after fixing the nit style comments. @huningxin do you want to review?

**Comment by DonnaWuDongxia** [comment]
@halton updated, thanks!

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObject.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    protected String mObjectId;
+    protected ExtensionInstanceHelper mInstanceHelper;
+    protected MessageHandler mHandler;
+
+    public BindingObject() {
+        mHandler = new MessageHandler();
+    }
+
+    public Object handleMessage(MessageInfo info) {
+       return mHandler.handleMessage(info);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObject.java]: nit, indent

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObject.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        mInstanceHelper = instance;
+    }
+
+    /*
+     * Called when this binding object is destroyed by JS.
+     */
+    public void onJsDestroyed() {}
+
+    /*
+     * Called when this object is binded with JS.
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObject.java]: binded -> bound

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectAutoJS.java [hunk 13]
```diff
@@ -10,32 +10,25 @@
 import org.json.JSONException;
 import org.json.JSONObject;
 
-public class XWalkExtensionBindingObject {
-    XWalkExtensionClient extensionClient;
-    int myObjectId;
-    int extInstanceId;
+public class BindingObjectAutoJS extends BindingObject{
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectAutoJS.java]: nit, space before `{`

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectStore.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    public Object onPostMessageToClass(MessageInfo info) {
+        Object result = null;
+        JSONArray args = info.getArgs();
+        try {
+            // args format:
+            // [memberName, [constructorJsName, memberArgs]]
+            MessageInfo newInfo = new MessageInfo(info);
+            String memberName = args.getString(0);
+            JSONArray originArgs = args.getJSONArray(1);
+            String cName = originArgs.getString(0);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectStore.java]: prefer full name as `constructorName` or `ctorName`

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectStore.java [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+            JSONArray memberArgs = originArgs.getJSONArray(1);
+
+            newInfo.setJsName(memberName);
+            newInfo.setArgs(memberArgs);
+
+            ReflectionHelper reflection =
+                info.getExtension().getTargetReflect(cName);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectStore.java]: indent

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectStore.java [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+                newInfo.setBinaryArgs(objectMethodArgs);
+            }
+            if (obj != null)
+                result = obj.handleMessage(newInfo);
+        } catch (JSONException e) {
+            Log.e(TAG, e.toString());
+        } catch (IndexOutOfBoundsException e) {
+            Log.e(TAG, e.toString());
+        } catch (NullPointerException e) {
+            Log.e(TAG, e.toString());
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/BindingObjectStore.java]: use || for above exceptions, I think current android sdk support that already.

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/EventTarget.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    public void dispatchEvent(String type) {
+        dispatchEvent(type, null);
+    }
+
+    public void dispatchEvent(String type, JSONObject data) {
+        try {
+            if (!mEvents.containsKey(type)) {
+                Log.w(TAG, "Attempt to dispatch to non-existing event :\n" + type);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/EventTarget.java]: I think `\n` should be replaced by space.

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/EventTarget.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+                return;
+            }
+
+            MessageInfo info = mEvents.get(type);
+            JSONArray args = new JSONArray();
+            if (data != null)
+                args.put(0, data);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/EventTarget.java]: prefer amend to oneline.

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/EventTarget.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+        } catch (JSONException e) {
+            Log.e(TAG, e.toString());
+        }
+    }
+
+    public void onAddEventListener(MessageInfo info) {
+        try {
+            String type = info.getArgs().getString(0);
+            if (mEvents.containsKey(type)) {
+                Log.w(TAG, "Trying to re-add the event :\n" + type);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/EventTarget.java]: ditto, `\n` -> space

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/EventTarget.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+        } catch (JSONException e) {
+            Log.e(TAG, e.toString());
+        }
+    }
+
+    public void onRemoveEventListener(MessageInfo info) {
+        try {
+            String type = info.getArgs().getString(0);
+            if (!mEvents.containsKey(type)) {
+                Log.w(TAG, "Attempt to remove non-existing event :\n" + type);
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/EventTarget.java]: ditto

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java [hunk 21]
```diff
@@ -32,13 +32,16 @@
...
     // The context used by extensions.
     protected XWalkExtensionContextClient mExtensionContext;
 
-    private Map<Integer, XWalkExtensionBindingObjectStore> instanceStores;
-
     // Reflection for JS stub generation
-    protected ReflectionHelper reflection;
+    protected ReflectionHelper mReflection;
+
+    protected boolean useJsStubGeneration;
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java]: rename with m prefix?

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java [hunk 22]
```diff
@@ -197,75 +209,28 @@ public void onBinaryMessage(int extensionInstanceID, byte[] message) {}
...
 
-    protected Object getBindingStore(int instanceId) {
-       return instanceStores.get(instanceId);
+    public MessageHandler getMessageHandler() {
+        return mHandler;
+    }
+    
+    public ReflectionHelper getTargetReflect(String cName) {
+        ReflectionHelper targetReflect = mReflection.getConstructorReflection(cName);
+        return (targetReflect != null)? targetReflect : mReflection;
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java]: space before `?`

File: extensions/renderer/xwalk_js_stub_wrapper.js [hunk 23]
```diff
@@ -122,45 +151,72 @@ function makeExtensionHelper(helper) {
...
+
+      if (msgObj.cmd == "invokeCallback") {
+        invokeMethod(msgObj.callbackId, msgObj.args);
+        return;
+      }
+      // Message about properties.
+      var targetHelper = this;
+      var msgType = MSG_TO_EXTENSION;
+
+      var cName = String(msgObj["constructorName"]).trim();
```
Comment (halton) [inline extensions/renderer/xwalk_js_stub_wrapper.js]: cName -> ctorName

**Comment by halton** [comment]
@DonnaWuDongxia I finished second round review, I think I can give lgtm after you fix them.

@huningxin do you want to review this PR?

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java (hunk 21 above)
Comment (DonnaWuDongxia) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java]: Yes, @halton

**Comment by DonnaWuDongxia** [comment]
@halton Updated!

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/MessageHandler.java [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+                Log.e(TAG, e.toString());
+            } catch (NoSuchMethodException e) {
+                Log.e(TAG, e.toString());
+            } catch (IllegalArgumentException e) {
+                Log.e(TAG, e.toString());
+            } catch (IllegalAccessException e) {
+                Log.e(TAG, e.toString());
+            } catch (InvocationTargetException e) {
+                Log.e(TAG, e.toString());
+            }
```
Comment (halton) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/MessageHandler.java]: Ditto, use `|`

**Comment by halton** [comment]
LGTM

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/JsApi.java [hunk 25]
```diff
@@ -28,4 +28,18 @@
      * This property is only meanful for functions.
      */
     public boolean withPromise() default false;
+
+    /*
+     * This property is only meanful for functions.
```
Comment (huningxin) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/JsApi.java]: typo "meaningful" -> "meanful"?

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/JsApi.java [hunk 26]
```diff
@@ -28,4 +28,18 @@
...
+     * It should be either the custom function name
+     * or just the function definition.
+     */
+    public String wrapArgs() default "";
+
+    /*
+     * This property is only meanful for functions.
```
Comment (huningxin) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/JsApi.java]: Ditto.

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/ExtensionInstanceHelper.java [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+    XWalkExtensionClient mExtension;
+    MessageHandler mHandler;
+    int mId;
+
+    public ExtensionInstanceHelper(XWalkExtensionClient extension, int id) {
+        mId = id;
+        mExtension = extension;
+        // Extend the handlers from extension object to each instance.
+        // For the case, that the developer register the handers by hand
+        // to the extenson itself.
```
Comment (huningxin) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/ExtensionInstanceHelper.java]: The last sentence of above comments is not very clear to me. "For the case", which case? "that" seems not needed. "register" would be "registers". "extension" is it the extension object or instance?

**Comment by huningxin** [comment]
Thanks for this PR! LGTM with some nits.

File: app/android/runtime_client/src/org/xwalk/app/runtime/extension/ExtensionInstanceHelper.java (hunk 27 above)
Comment (DonnaWuDongxia) [inline app/android/runtime_client/src/org/xwalk/app/runtime/extension/ExtensionInstanceHelper.java]: Thanks for comments. I changed the words. It means copying all the handlers registered to the extension object to each instance, as we have only extension object exposed to developer currently.

**Comment by DonnaWuDongxia** [comment]
Thanks @halton @huningxin , updated.

**Comment by huningxin** [comment]
Code LGTM. Thanks for the update!

**Comment by halton** [comment]
Will merge when trybot green.

**Pull request merged**